### PR TITLE
Update to use PSR-11 for plugin managers

### PR DESCRIPTION
**Pull request by weierophinney**
This patch updates zend-config to typehint on PSR-11 for each of its plugin managers, and provides standalone versions that have no dependencies other than the PSR-11 interfaces; this allows usage of the `Factory` without having zend-servicemanager installed.

Additionally, this patch does the following:

- Updates the minimum supported versions of dependencies to the latest releases of each. If both v2 and v3 versions exist, it allows either.
- Updates the JSON reader and writer to allow usage with ext/json, falling back to zend-json if present. This allows usage without zend-json being present.
- Adds malukehno/docheader as a development requirement, in order to enforce license headers in all class files.

### CODE REVIEW

File: src/Reader/Json.php [hunk 1]
```diff
@@ -102,4 +92,79 @@ protected function process(array $data)
...
+     *
+     * @param string $data
+     * @return array
+     * @throws Exception\RuntimeException for any decoding errors.
+     * @throws Exception\MissingExtensionException if neither ext/json nor
+     *     zend-json are available.
+     */
+    private function decode($data)
+    {
+        if (function_exists('json_decode')) {
```
Comment (webimpress) [inline src/Reader/Json.php]: From PHP docs:
> As of PHP 5.2.0, the JSON extension is bundled and compiled into PHP by default.
http://php.net/manual/en/json.installation.php

File: src/StandaloneWriterPluginManager.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+class StandaloneWriterPluginManager implements ContainerInterface
+{
+    private $knownPlugins = [
+        'ini'            => Writer\Ini::class,
+        'json'           => Writer\Json::class,
+        'php'            => Writer\PhpArray::class,
+        'phparray'       => Writer\PhpArray::class,
+        'xml'            => Writer\Xml::class,
+        'yaml'           => Writer\Yaml::class,
```
Comment (webimpress) [inline src/StandaloneWriterPluginManager.php]: Can we have here less spaces before `=>`, please?

File: src/Writer/Json.php [hunk 3]
```diff
@@ -18,9 +17,22 @@ class Json extends AbstractWriter
      *
      * @param  array $config
      * @return string
+     * @throws Exception\MissingExtensionException if neither ext/json nor
+     *     zendframework/zend-json are available.
      */
     public function processConfig(array $config)
     {
-        return JsonFormat::encode($config);
+        if (function_exists('json_encode')) {
```
Comment (webimpress) [inline src/Writer/Json.php]: As noted above, probably we don't need it.

File: test/StandaloneReaderPluginManagerTest.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            'yaml'           => ['yaml', Reader\Yaml::class],
+            'YAML'           => ['YAML', Reader\Yaml::class],
+            'javaproperties' => ['javaproperties', Reader\JavaProperties::class],
+            'javaProperties' => ['javaProperties', Reader\JavaProperties::class],
+            'JAVAPROPERTIES' => ['JAVAPROPERTIES', Reader\JavaProperties::class],
+        ];
+    }
+
+    /**
+     * @dataProvider supportedConfigExtensions
```
Comment (webimpress) [inline test/StandaloneReaderPluginManagerTest.php]: Can we add here also `@param TYPE NAME` for method parameters, please?

```php
* @param string $extension
* @param string $expectedType
```

File: test/StandaloneReaderPluginManagerTest.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            Reader\Ini::class            => [Reader\Ini::class],
+            Reader\Json::class           => [Reader\Json::class],
+            Reader\Xml::class            => [Reader\Xml::class],
+            Reader\Yaml::class           => [Reader\Yaml::class],
+            Reader\JavaProperties::class => [Reader\JavaProperties::class],
+        ];
+    }
+
+    /**
+     * @dataProvider supportedConfigClassNames
```
Comment (webimpress) [inline test/StandaloneReaderPluginManagerTest.php]: Can we add here:
```php
* @param string $class
```
please?

File: test/StandaloneWriterPluginManagerTest.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            'PhpArray' => ['PhpArray', Writer\PhpArray::class],
+            'xml'      => ['xml', Writer\Xml::class],
+            'XML'      => ['XML', Writer\Xml::class],
+            'yaml'     => ['yaml', Writer\Yaml::class],
+            'YAML'     => ['YAML', Writer\Yaml::class],
+        ];
+    }
+
+    /**
+     * @dataProvider supportedConfigTypes
```
Comment (webimpress) [inline test/StandaloneWriterPluginManagerTest.php]: As above, can we update PHPDocs here, please?

File: test/StandaloneWriterPluginManagerTest.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            Writer\Ini::class      => [Writer\Ini::class],
+            Writer\Json::class     => [Writer\Json::class],
+            Writer\PhpArray::class => [Writer\PhpArray::class],
+            Writer\Xml::class      => [Writer\Xml::class],
+            Writer\Yaml::class     => [Writer\Yaml::class],
+        ];
+    }
+
+    /**
+     * @dataProvider supportedConfigClassNames
```
Comment (webimpress) [inline test/StandaloneWriterPluginManagerTest.php]: Update PHPDocs please.

File: src/Factory.php [hunk 8]
```diff
@@ -204,47 +203,47 @@ public static function toFile($filename, $config)
...
+    public static function setReaderPluginManager(ContainerInterface $readers)
     {
         static::$readers = $readers;
     }
 
     /**
      * Get the reader plugin manager
      *
-     * @return ReaderPluginManager
+     * @return ContainerInterface
```
Comment (webimpress) [inline src/Factory.php]: Maybe we should have here also `StandaloneReaderPluginManager`, so generic one and specific/default?

Yes, I know that `StandaloneReaderPluginManager implements ContainerInterface`.

File: src/Factory.php [hunk 9]
```diff
@@ -204,47 +203,47 @@ public static function toFile($filename, $config)
...
+    public static function setWriterPluginManager(ContainerInterface $writers)
     {
         static::$writers = $writers;
     }
 
     /**
      * Get the writer plugin manager
      *
-     * @return WriterPluginManager
+     * @return ContainerInterface
```
Comment (webimpress) [inline src/Factory.php]: As above, maybe we need to have here `ContainerInterface|StandaloneWriterPluginManager`?

File: src/Reader/Json.php (hunk 1 above)
Comment (weierophinney) [inline src/Reader/Json.php]: Yes, but many distributions still ship it as an optional package. As such, we still need to verify, and have zend-json as a polyfill.

File: src/Factory.php (hunk 8 above)
Comment (weierophinney) [inline src/Factory.php]: The whole point with this change was to signal that you can provide generic implementations based on `ContainerInterface`.

I'll update the method description to indicate the default implementation, however.

File: test/StandaloneReaderPluginManagerTest.php (hunk 4 above)
Comment (weierophinney) [inline test/StandaloneReaderPluginManagerTest.php]: Done.

File: test/StandaloneReaderPluginManagerTest.php (hunk 5 above)
Comment (weierophinney) [inline test/StandaloneReaderPluginManagerTest.php]: Done.

File: test/StandaloneWriterPluginManagerTest.php (hunk 6 above)
Comment (weierophinney) [inline test/StandaloneWriterPluginManagerTest.php]: Done.

File: test/StandaloneWriterPluginManagerTest.php (hunk 7 above)
Comment (weierophinney) [inline test/StandaloneWriterPluginManagerTest.php]: Done.

File: src/Reader/Json.php (hunk 1 above)
Comment (Ocramius) [inline src/Reader/Json.php]: `"require":{"ext-json":"*"}` ;-)

Comment (weierophinney) [inline src/Reader/Json.php]: That's even easier... :smile:

Comment (weierophinney) [inline src/Reader/Json.php]: Updated to remove zend-json entirely and require ext-json via Composer.

Comment (webimpress) [inline src/Reader/Json.php]: Ok, I like it, but now we lost zend-json fallback IF somebody doesn't have this extension and can't add it (_IF_ it is even possible to not have it now on PHP 5.6+)

Comment (Ocramius) [inline src/Reader/Json.php]: @webimpress seriously: their loss. A system without json support in 2017? :|

**Pull request merged**